externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -191,4 +191,92 @@ def Tcp_BroadcastOp : Tcp_Op<"broadcast", [
   let hasVerifier = 1;
 }
 
+def Tcp_YieldOp : Tcp_Op<"yield", [Terminator, Pure]> {
+  let summary = "Yields values from regions in Tcp";
+
+  let description = [{
+    Terminates and yields values from regions in Tcp.
+
+    The values that are yielded should correspond to the parent operation's
+    results.
+  }];
+
+  let arguments = (ins
+    Variadic<Tcp_TensorOrScalar>:$ins
+  );
+
+  let builders = [
+    OpBuilder<(ins), [{ }]>,
+  ];
+
+  let assemblyFormat = "$ins attr-dict `:` type($ins)";
+}
+
+def Tcp_GroupOp : Tcp_Op<"group", [
+          AffineScope,
+          AutomaticAllocationScope,
+          ParentOneOf<["func::FuncOp", "tcp::GroupOp", "tcp::IsolatedGroupOp"]>,
```

Comment:
Groups can be nested? Also they can't be in control flow?

Reply:
Yes, they can be nested. 

Good question regarding control flow. Do you see any problems with groups inside control flow? I didn't add them here because we don't have them yet in Tcp.

Reply:
My view of cflow and groups is that they should be ok because groups are not a runtime artifact but a compile-time aggregation for the purposes of aiding codegen.

Reply:
How about control flow within groups?

Reply:
Can you explain more why groups are not runtime artifact? I am guessing they correspond to some unit of work that is somehow "launched".

Reply:
> How about control flow within groups?

Both grouping ops will not have control flow within them. The reasoning behind that is just to keep the transformations to move ops into and out of groups simple. Just clarified this in the description. 

Do you think we need to support control flow in groups?

Reply:
> Can you explain more why groups are not runtime artifact?

Grouping ops in Tcp are meant to facilitate lowering efficiently to different backends. We do not foresee these grouping ops to be present once we lower from Tcp to a backend dialect.

Reply:
> Do you think we need to support control flow in groups?

No explicit control flow.  I don't see how we can prevent "implicit" control flow though -- a functional `while` loop for instance -- as long as we have allow non-TCP operations in groups.

We should have a verifier check for this.

Reply:
Your point regarding implicit control flow is valid. But how do you propose checking that the group region does not include any such ops? I don't see them implementing any trait / interface.

Reply:
> > Can you explain more why groups are not runtime artifact?
> 
> Grouping ops in Tcp are meant to facilitate lowering efficiently to different backends. We do not foresee these grouping ops to be present once we lower from Tcp to a backend dialect.

Well, the grouping gets lowered somehow to the backend. I think traits like `AutomaticAllocationScope` etc. are related to runtime (someone needs to do the allocation). So my sense is that there is some bearing on runtime hand-shake here that isnt accounted for.

Reply:
@MaheshRavishankar Just replied to your comment in the doc: https://docs.google.com/document/d/1jHXI-ZSyt-3nFlO0ZtQZX0E41k4mCmuO7djmrvpyHms/edit?disco=AAAAq40iWww

Reply:
Regarding verifying for control flow ops, clarified with @sanjoy offline. By construction these ops only allow one basic block (because their body is a `SizedRegion<1>`). There is no other feasible way to catch all control flow ops AFAICT.